apollo-router/src/cache/mod.rs

```diff
@@ -0,0 +1,250 @@
+        let mut locked_wait_map = self.wait_map.lock().await;
+        match locked_wait_map.get(key) {
+            Some(waiter) => {
+                // Register interest in key
+                let receiver = waiter.subscribe();
+                Entry {
+                    inner: EntryInner::Receiver { receiver },
+                }
+            }
+            None => {
+                let (sender, _receiver) = broadcast::channel(1);
+
+                locked_wait_map.insert(key.clone(), sender.clone());
+
+                // we must not hold a lock over the wait map while we are waiting for a value from the
+                // cache. This way, other tasks can come and register interest in the same key, or
+                // request other keys independently
+                drop(locked_wait_map);
+
+                if let Some(value) = self.storage.get(key).await {
+                    let mut locked_wait_map = self.wait_map.lock().await;
+                    let _ = locked_wait_map.remove(key);
+                    let _ = sender.send(value.clone());
+
+                    return Entry {
+                        inner: EntryInner::Value(value),
+                    };
+                }
+
+                let k = key.clone();
+                // when _drop_signal is dropped, either by getting out of the block, returning
+                // the error from ready_oneshot or by cancellation, the drop_sentinel future will
+                // return with Err(), then we remove the entry from the wait map
+                let (_drop_signal, drop_sentinel) = oneshot::channel::<()>();
+                let wait_map = self.wait_map.clone();
+                tokio::task::spawn(async move {
+                    let _ = drop_sentinel.await;
+                    let mut locked_wait_map = wait_map.lock().await;
+                    let _ = locked_wait_map.remove(&k);
+                });
+
+                Entry {
+                    inner: EntryInner::First {
+                        sender,
+                        key: key.clone(),
+                        cache: self.clone(),
+                        _drop_signal,
+                    },
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn insert(&self, key: K, value: V) {
+        self.storage.insert(key, value.clone()).await;
+    }
+
+    pub(crate) async fn remove_wait(&self, key: &K) {
+        let mut locked_wait_map = self.wait_map.lock().await;
+        let _ = locked_wait_map.remove(key);
+    }
+}
+
+pub(crate) struct Entry<K: Clone + Send + Eq + Hash, V: Clone + Send> {
+    inner: EntryInner<K, V>,
+}
+enum EntryInner<K: Clone + Send + Eq + Hash, V: Clone + Send> {
+    First {
+        key: K,
+        sender: broadcast::Sender<V>,
+        cache: DeduplicatingCache<K, V>,
+        _drop_signal: oneshot::Sender<()>,
+    },
+    Receiver {
+        receiver: broadcast::Receiver<V>,
+    },
+    Value(V),
+}
```

Comment:
I think this happened because you don't want to provide a callback to cache::get(), I wonder if we could make it leaner

Reply:
having everything in the `Entry` structure is an attempt at a safer API, that will automatically use the drop signal or drop the receiver without having to go through the cache structure again

Reply:
I see, It's a tuff tradeoff since we end up with cache.get().await.get().await which feels a bit odd

Reply:
we have this in APQ because that layer will not generate the value on cache miss. The more general case would be to match on `cache.get().await` then decide if we await on the entry or not